File: 397_Integer_Replacement/main.py
class Solution:
    def integerReplacement(self, n: int) -> int:
        @cache
        def f(x):
            if x==1:
                return 0
            if x%2==0:
                return 1+f(x/2)
            else:
                return 1+min(f(x+1),f(x-1))

        return f(n)

File: 397_Integer_Replacement/main.go
func integerReplacement(n int) int {
	cnt := 0
	for n != 1 {
		if n%2 == 0 {
			n = n / 2
		} else {
			if n == 3 || n%4 == 1 {
				n -= 1
			} else {
				n += 1
			}
		}
		cnt++
	}

	return cnt
}
